File: practicas/practica4_rb/lib/SpaceCity.rb
#encoding: utf-8

require_relative 'SpaceStation'
require_relative 'SpaceCityToUI'
require_relative 'ShotResult'
require_relative 'Transformation'

module Deepspace

  class SpaceCity < SpaceStation

    def initialize(base, rest)
      fromOther base
      @base = base
      @collaborators = rest
    end

    def collaborators
      @collaborators.clone
    end

    def protection
      protection_power = 0
      protection_power += @base.protection

      @collaborators.each do |c|
        protection_power += c.protection
      end
      protection_power
    end

    def fire
      fire_power = 0
      fire_power += @base.fire

      @collaborators.each do |c|
        fire_power += c.fire
      end

      fire_power
    end

    def setLoot(l)
      super l
      Transformation::NOTRANSFORM
    end

    def getUIversion
      SpaceCityToUI.new self
    end

  end
end

File: practicas/practica4/src/main/java/deepspace/SpaceCity.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author inmagalvez
 */
public class SpaceCity extends SpaceStation {
    private ArrayList<SpaceStation> collaborators;
    private SpaceStation base;
    
    //Preguntar si hay que añadir base a collaborators y si hay que inicializar base
    SpaceCity(SpaceStation base, ArrayList<SpaceStation> rest){
        super(base);
        collaborators = rest;
        //collaborators.add(0, base);
        this.base = base;
    }
    
    ArrayList<SpaceStation> getCollaborators(){
        return new ArrayList<>(collaborators);
    }
    
    @Override
    public float fire(){
        float res = 0.0f;
        for(SpaceStation i : collaborators){
            res += i.fire();
        }
        return res;
    }
    
    @Override
    public float protection(){
        float res = 0.0f;
        for(SpaceStation i : collaborators){
            res += i.protection();
        }
        return res;
    }
    
    @Override
    public Transformation setLoot(Loot loot){
        super.setLoot(loot);
        return Transformation.NOTRANSFORM;
    }
    
    @Override
    public SpaceStationToUI getUIversion() {
        return new SpaceCityToUI(this);
    }

}
